### ISSUE TITLE
Wish: Also call IClientStartup on DLLs which don't contain the word Oqtane

### ISSUE THREAD

**Issue by iJungleboy** [oqtane/oqtane.framework#2394]
Current behavior is that any DLL which has `IClientStartup` will only be initialized if the DLL contains the word `Oqtane`.

## The `Oqtane` in your own Namespace-Problem
This is quite a big problem, because if the namespace of your own modules contains `Oqtane` the code will run into many issues. So if your module is say `ToSic.Oqtane.Something.Client` then any code that accesses the `Oqtane` namespace will need to be super specific, because the compiler will first check the own namespace. 

Because of this we decided to go with a naming like `ToSic.Oqt.Something.Client` - I believe there is no better way around the "common `Oqtane` in Namespaces" problem. 

But this again leads to a new problem: Your DLLs won't trigger `IClientStartup`. IMHO this doesn't make sense, since it's a very specific interface and if it exists, it should be called. In the end it forces us to name the DLL differently than the namespace - like `ToSic.Oqt.Something.Client.Oqtane.dll` just to make sure the services will still be initialized. 

So please change the behavior to always call IClientStartup, even if the assembly doesn't contain `Oqtane` in the name.

**Comment by sbwalker** [comment]
This topic has been covered many times previously... please search prior Issues and Discussions. This behavior will not be changed for the reasons outlined previously.

**Comment by sbwalker** [comment]
In a nutshell the current behavior favors performance over convenience (as performance is the overarching Oqtane theme). 

Rather than iterating through every assembly in the /bin, loading them into the app domain, and then interrogating them for interfaces via reflection... Oqtane relies on a file naming convention to uniquely identify only those assemblies which are relevant. This provides a significant performance benefit at startup - which is an important scenario, especially during development when you are starting and stopping the app regularly. 

That being said, with all of the performance improvements in .NET Core in recent releases it would be worthwhile to capture some metrics of the actual time it takes to do the brute force approach vs the selective approach. It used to be a 5-10 second difference.

**Comment by iJungleboy** [comment]
@sbwalker ok thanks. The performance issue would be important, I totally agree. 

I've tried to read the code and this is my understanding. We seem to have multiple places where these checks happen:

1. **Oqtane.Client** - `RegisterClientStartups` - this will `LoadClientAssemblies` from the backend and put them in memory on the client. IMHO the extra check here wouldn't be necessary, since the check already happens in the backend and restricts that only such DLLs actually get delivered to the client. 
2. **Oqtane.Server** - `OqtaneServiceCollectionExtensions.AddOqtaneServices(...)` which to me appears to do the same for ServerRendering with with slightly different code. I assume it's for server-side-rendering and should of course have comparable behavior to the client-side rendering. Here speed could matter, but it should result in the same behavior / same DLLs being loaded as with the client.
3. **Oqtane.Server** - `InstallationController` which prepares the DLLs for the browser in client-side processing.

This is what IMHO the `InstallationController` and the `AddOqtaneServices` seem to do:

1. The server runs some code in the `InstallationController.cs` - endpoint `/load` -> `GetAssemblies()` to find all DLLs which it should pass to the client
4. It will get all assemblies containing `oqtane` and having an `IModule...`, `IThemeControl` or `IClientStartup` 
5. It will then loop through lists of `Dependencies` on `Module`/`Theme` to see if it needs anything else to load. from what I can see, these `Dependencies` would have to be maintained manually, there is no automatic-find-dependencies or something
6. It will then zip all the DLLs into one and pass it to the client

If my understanding of this is correct, then there are a few obvious things we must achieve

1. We shouldn't include all DLLs - but only DLLs which really have code for the client - so obviously code containing `IModule`, `IClientStartup` etc. and ideally any supporting DLLs required. 
2. As of now, this is "sped up" by pre-filtering DLLs which contain the word `oqtane` - so that many DLLs don't need looking into. 

This leaves the question: How much time to we save (as of today). 

I ran some quick tests using the `Stopwatch` on the `AddOqtaneServices` (server side rendering). 

* With the existing implementation, a few runs result in a hit (on my system) of ca. 0.105 seconds (105ms)
* Replacing `GetOqtaneAssemblies()` with `GetAssemblies()` increased that to about 0.505 seconds (505ms) on average - so + 0.4 seconds

There may be some room for optimizations, and I could imagine that the client-side stack only may have different timing results because it does other things. But these just my quick findings. 

So basically I understand your thoughts and there is a difference, which could be significant or not, this is up to you to judge. 

Feel free to close this or keep it open, whatever you think is best. thanks 😀

**Comment by sbwalker** [comment]
Your research identified some of the areas related to assembly, but not the primary one. During Server Startup there is a call to services.AddOqtane(). This method is responsible for loading Oqtane assemblies from the /bin into the App Domain. You can read this blog to understand why this is necessary https://www.oqtane.org/blog/!/11/assembly-loading-in-blazor-and--net-core/. So it is this method which evaluates the file names of assemblies to determine whether or not they need to be loaded into the App Domain. Note that once an assembly is loaded into the App Domain it cannot be unloaded so you need to be very selective with what you load - as every assembly loaded consumes server RAM. However in order to interrogate an assembly to determine if they implement an interface, the assembly first needs to be loaded into the App Domain. So its a catch-22... and to deal with this challenge Oqtane uses a white-list approach to identify assemblies by file name. So although your performance tests above are interesting, they do not measure the actual workload in question. That being said I am surprised at the results, as the test you ran basically shows the difference of iterating through the entire collection of assemblies already loaded into the App Domain versus using a selective approach based on assembly name - and the difference is almost a half second. I would have assumed that in-memory processing would be really fast and would not have benefited from the optimization this much.

**Comment by iJungleboy** [comment]
My guess is that the time used here is just used earlier. So my guess is that the time I measured caused assemblies to load which would otherwise have been loaded later. Because I'm 100% sure that if it were already in memory, it would be much, much faster.

But I understand your concern and my wish isn't so big that it needs further thought. I'll close this for now.

Thanks for the constructive conversation 😀
